libraries/AP_UAVCAN/AP_UAVCAN_SLCAN.cpp

```diff
@@ -0,0 +1,602 @@
+ * for Zubax Babel
+ */
+
+#include <AP_HAL/AP_HAL.h>
+
+#if HAL_WITH_UAVCAN && !HAL_MINIMIZE_FEATURES && CONFIG_HAL_BOARD == HAL_BOARD_CHIBIOS
+
+#include "AP_UAVCAN_SLCAN.h"
+#include <AP_SerialManager/AP_SerialManager.h>
+#include <AP_Common/Semaphore.h>
+
+#include <AP_HAL_ChibiOS/CANSerialRouter.h>
+
+extern const AP_HAL::HAL& hal;
+
+static uint8_t nibble2hex(uint8_t x)
+{
+    // Allocating in RAM because it's faster
+    static uint8_t ConversionTable[] =
+    {
+        '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'A', 'B', 'C', 'D', 'E', 'F'
+    };
+    return ConversionTable[x & 0x0F];
+}
+
+static bool hex2nibble_error;
+
+static uint8_t hex2nibble(char ch)
+{
+    // Must go into RAM, not flash, because flash is slow
+    static uint8_t ConversionTable[] =
+    {
+        255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
+        255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
+        255, 255, 255, 255,
+        0, 1, 2, 3, 4, 5, 6, 7, 8, 9,           // 0..9
+        255, 255, 255, 255, 255, 255, 255,
+        10, 11, 12, 13, 14, 15,                 // A..F
+        255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
+        255, 255, 255, 255,
+        10, 11, 12, 13, 14, 15,                 // a..f
+        255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
+        255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
+        255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
+        255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
+        255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
+        255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
+        255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255
+    };
+
+    const uint8_t out = ConversionTable[int(ch)];
+    if (out == 255)
+    {
+        hex2nibble_error = true;
+    }
+    return out;
+}
+
+
+bool SLCAN::CAN::push_Frame(uavcan::CanFrame &frame)
+{
+    SLCAN::CanRxItem frm;
+    frm.frame = frame;
+    frm.flags = 0;
+    frm.utc_usec = AP_HAL::micros64();
+    slcan_router().route_frame_to_can(frm.frame, frm.utc_usec);
+    return rx_queue_.push(frm);
+}
+
+/**
+ * General frame format:
+ *  <type> <id> <dlc> <data>
+ * The emitting functions below are highly optimized for speed.
+ */
+bool SLCAN::CAN::handle_FrameDataExt(const char* cmd)
+{
+    uavcan::CanFrame f;
+    hex2nibble_error = false;
+    f.id = f.FlagEFF |
```

Comment:
I moved most of the parsing code directly from babel code by Pavel. If you think there would be a gain in using sscanf rather than this method, I can give it a shot.